scaaml/capture/capture_runner.py

```diff
@@ -0,0 +1,141 @@
+        capture).
+
+        Args:
+          crypto_algorithms: Provide attack points and information about the
+            encryption algorithm. A separate capture is run for each element.
+          communication: Object that communicates with the observed chip.
+          control: Control of the target board.
+          scope: Scope which does the measurements.
+          dataset: Dataset to save measurements into.
+
+        Typical usage example:
+          capture_runner = CaptureRunner(crypto_algorithms=[crypto_alg],
+                                         scope=scope,
+                                         communication=target,
+                                         control=control,
+                                         dataset=dataset)
+          # Continue capturing the dataset.
+          capture_runner.capture()
+        """
+        self._crypto_algorithms = crypto_algorithms
+        self._communication = communication
+        self._control = control
+        self._scope = scope
+        self._dataset = dataset
+
+    @abstractmethod
+    def get_crypto_input(self, kt_element) -> AbstractCryptoInput:
+        """Process single element from ResumeKTI and return information for
+        the crypto algorithm.
+
+        Args:
+          kt_element: Single element received from looping over ResumeKTI
+            instance. A pair of np arrays.
+
+        Returns: An instance of input for cryptographic algorithm (for example
+          an object holding a key and plaintext).
+        """
+
+    @abstractmethod
+    def get_attack_points_and_measurement(
+            self, crypto_alg: AbstractSCryptoAlgorithm,
+            crypto_input: AbstractCryptoInput) -> Tuple[Dict, Dict]:
+        """Get attack points and measurement. Repeat capture if necessary.
+        Raises if hardware fails.
+
+        Args:
+          crypto_alg: The object used to get attack points.
+          crypto_input: The input from ResumeKTI (a pair of np arrays).
+
+        Returns: Attack points and physical measurement. These are to be used
+          directly by scaaml.io.Dataset.write_example.
+
+        Raises: If capturing failed in an unrecoverable way.
+        """
+
+    def _stabilize_capture(self, crypto_alg: AbstractSCryptoAlgorithm):
+        """Stabilize the capture by capturing a few traces.
+
+        Args:
+          crypto_alg: The object used to get stabilization attack points.
+        """
+        skti = crypto_alg.get_stabilization_kti()
+        crypto_input = self.get_crypto_input(next(skti))
+        try:
+            _, _ = self.get_attack_points_and_measurement(
+                crypto_alg=crypto_alg, crypto_input=crypto_input)
+        except (Warning, OSError, AssertionError):
+            pass
+        # Stabilize the capture
+        for _ in range(10):
+            _, _ = self.get_attack_points_and_measurement(
+                crypto_alg=crypto_alg, crypto_input=crypto_input)
+
+    def capture(self):
```

Comment:
I know we already discussed it but from the capture point of view, the fact that ML requires 2 datasets for training should be transparent.
If we capture `training` the tooling should take care of capturing (and eventually resuming in case of interruption) `train` and `test` subsets all at once. The geometry of `test` should be inferred from the requested one for `train`.

Reply:
I thought this was going to be another PR. But I am happy to do it here :)